porch/func/internal/podevaluator.go

```diff
@@ -0,0 +1,212 @@
+	ref, err := name.ParseReference(req.Image)
+	if err != nil {
+		return nil, err
+	}
+	img, err := remote.Image(ref, remote.WithAuthFromKeychain(authn.DefaultKeychain))
+	if err != nil {
+		return nil, err
+	}
+	cf, err := img.ConfigFile()
+	cfg := cf.Config
+	var fnCmd []string
+	if len(cfg.Entrypoint) != 0 {
+		fnCmd = cfg.Entrypoint
+	} else {
+		fnCmd = cfg.Cmd
+	}
+
+	cmd := append([]string{
+		path.Join(volumeMountPath, "wrapper-server"),
+		"--port", strconv.Itoa(defaultWrapperServerPort), "--",
+	}, fnCmd...)
+
+	// Create a pod
+	pod := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Pod",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:    pe.namespace,
+			GenerateName: "krm-fn-",
+			// TODO: add labels
+		},
+		Spec: corev1.PodSpec{
+			InitContainers: []corev1.Container{
+				{
+					Name:  "copy-wrapper-server",
+					Image: pe.wrapperServerImage,
+					Command: []string{
+						"sh", "-c",
+						fmt.Sprintf("cp /wrapper-server/* %v", volumeMountPath),
+					},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "wrapper-server-tools",
+							MountPath: volumeMountPath,
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name:  "function",
+					Image: req.Image,
+					Command: []string{
+						"sh", "-c",
+						strings.Join(cmd, " "),
+					},
+					ReadinessProbe: &corev1.Probe{
+						ProbeHandler: corev1.ProbeHandler{
+							Exec: &corev1.ExecAction{
+								Command: []string{
+									"/wrapper-server-tools/grpc-health-probe",
+									"-addr", fmt.Sprintf("localhost:%v", strconv.Itoa(defaultWrapperServerPort)),
+								},
+							},
+						},
+					},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "wrapper-server-tools",
+							MountPath: "/wrapper-server-tools",
+						},
+					},
+				},
+			},
+			Volumes: []corev1.Volume{
+				{
+					Name: "wrapper-server-tools",
+					VolumeSource: corev1.VolumeSource{
+						EmptyDir: &corev1.EmptyDirVolumeSource{},
+					},
+				},
+			},
+		},
+	}
+	err = pe.kubeClient.Create(ctx, pod)
+	if err != nil {
+		return nil, err
+	}
+	klog.Infof("created KRM function evaluator pod %v", pod.Name)
+
+	defer func() {
+		// TODO: keep the pod around for some time for potential reuse.
+		klog.Infof("deleting KRM function evaluator pod %v", pod.Name)
+		if err := pe.kubeClient.Delete(ctx, pod); err != nil {
+			klog.Warningf("failed to delete KRM function evaluator pod %v: %v", pod.Name, err)
+		}
+	}()
+
+	// Wait until the pod is Running
+	if err = wait.PollImmediate(100*time.Millisecond, 60*time.Second, func() (done bool, err error) {
+		err = pe.kubeClient.Get(ctx, client.ObjectKeyFromObject(pod), pod)
+		if err != nil {
+			return false, err
+		}
+		if pod.Status.Phase != "Running" {
+			return false, nil
+		}
+		for _, cond := range pod.Status.Conditions {
+			if cond.Type == corev1.PodReady {
```

Comment:
does the gRPC liveness probe affect pod readiness?

Reply:
No. They are independent probes.

If a liveness probe keeps failing for some time, it will be recreated. But it doesn't affect the readiness.

Reply:
We should add liveness probe when we make it long lived.